agent/testagent.go

```diff
@@ -296,7 +296,11 @@ func (a *TestAgent) Shutdown() error {
 
 	// shutdown agent before endpoints
 	defer a.Agent.ShutdownEndpoints()
-	return a.Agent.ShutdownAgent()
+	if err := a.Agent.ShutdownAgent(); err != nil {
+		return err
+	}
+	<-a.Agent.ShutdownCh()
```

Comment:
isn't this going to introduce latency before being able to start the next test? what was your thought behind it? 💛

Reply:
Deliberate latency to stem possible test bleedover leading to flakiness. Also required so that the changes over in https://github.com/hashicorp/consul/pull/6425 actually have a chance of truly working (yay shared ports resource!)

Reply:
Would it be possible to not block the return on releasing the port, under the assumption that other, not-currently-in-use, ports should be available while that one remains busy? (Or am I misunderstanding #6425 conceptually?)

Reply:
All resources allocated in a test function should be released before the test function should be allowed to return. That includes stray temp files, ports opened, and goroutines launched.

In the case of a full agent launch there is a lot going on in there. The only way to guarantee that it's done doing whatever it's doing is to wait for it to be fully shutdown.

#6425 adds take/return semantics around all test-related port requests, but it's on the honor system (because there is no alternative). When a deferred call of `freeport.Return` actually runs in order for the honor system to hold _nothing can be using that port anymore_.

This includes ports that are explicitly bound, any any ports that the test reserved for test reasons, like setting up two agents, joining them, and then explicitly shutting down one of them to verify that the first agent reacts well. In that case the ports for the second agent aren't _bound_ anymore *but* no other test can safely use them yet because the first agent is still expecting those ports to no longer be in use.

This PR (#6441) isn't explicitly about the freeport issue, but the more general dirty behavior of "tests don't wait until they are done running to return control".

Reply:
SGTM. Tests aren't taking any longer than normal anyway, or not noticeably so.